lib/ansible/module_utils/basic.py

```diff
@@ -920,6 +924,20 @@ def __init__(self, argument_spec, bypass_checks=False, no_log=False,
                            ' Update the code for this module  In the future, AnsibleModule will'
                            ' always check for invalid arguments.', version='2.9')
 
+    @property
+    def tmpdir(self):
+        # if _ansible_tmpdir was not set, the module needs to create it and
+        # clean it up once finished.
+        if self._tmpdir is None:
+            basedir = os.path.expanduser(self._remote_tmp)
```

Comment:
We should do expandvars here as well: ```expanduser(expandvars(self._remote_tmp))```.  People might use remote_tmp="$HOME", for instance.